diff --git a/lib/src/socketBase.js b/lib/src/socketBase.js
index 17eafe729215cc5295a76f8ff7316ea24a161b16..27f1e9527b19e449ca84f123937820cb92747677 100644
--- a/lib/src/socketBase.js
+++ b/lib/src/socketBase.js
@@ -31,18 +31,32 @@ var SocketBase = /** @class */ (function (_super) {
         _this.endpointTerminated = _this.endpointTerminated.bind(_this);
         _this.xrecv = _this.xrecv.bind(_this);
         _this.hiccuped = _this.hiccuped.bind(_this);
+        _this.endpointAttached = function (endpoint) {
+            _this.attachEndpoint(endpoint);
+            _this.emit('ready', endpoint);
+        };
+        _this.endpointLost = function (endpoint) {
+            _this.emit('lost', endpoint);
+        };
+        _this.endpointTerminatedInternal = function (endpoint) {
+            _this.endpointTerminated(endpoint);
+            _this.emit('lost', endpoint);
+        };
+        _this.endpointRecovered = function (endpoint) {
+            _this.hiccuped(endpoint);
+            _this.emit('ready', endpoint);
+        };
         return _this;
     }
     SocketBase.prototype.connect = function (address) {
         if (address.startsWith("ws://") || address.startsWith("wss://")) {
             var endpoint = new webSocketEndpoint_1.default(address, this.options);
-            endpoint.on('attach', this.attachEndpoint);
-            endpoint.on('terminated', this.endpointTerminated);
+            endpoint.on('attach', this.endpointAttached);
+            endpoint.on('lost', this.endpointLost);
+            endpoint.on('terminated', this.endpointTerminatedInternal);
             endpoint.on('message', this.xrecv);
-            endpoint.on('hiccuped', this.hiccuped);
+            endpoint.on('hiccuped', this.endpointRecovered);
             this.endpoints.push(endpoint);
-            if (!this.options.immediate)
-                this.attachEndpoint(endpoint);
         }
         else {
             throw new Error('unsupported transport');
@@ -51,13 +65,15 @@ var SocketBase = /** @class */ (function (_super) {
     SocketBase.prototype.disconnect = function (address) {
         var endpoint = lodash_1.find(this.endpoints, function (e) { return e.address === address; });
         if (endpoint) {
-            endpoint.removeListener('attach', this.attachEndpoint);
-            endpoint.removeListener('terminated', this.endpointTerminated);
+            endpoint.removeListener('attach', this.endpointAttached);
+            endpoint.removeListener('lost', this.endpointLost);
+            endpoint.removeListener('terminated', this.endpointTerminatedInternal);
             endpoint.removeListener('message', this.xrecv);
-            endpoint.removeListener('hiccuped', this.hiccuped);
+            endpoint.removeListener('hiccuped', this.endpointRecovered);
             endpoint.close();
             lodash_1.pull(this.endpoints, endpoint);
             this.endpointTerminated(endpoint);
+            this.emit('lost', endpoint);
         }
     };
     SocketBase.prototype.bind = function (address, server) {
@@ -89,13 +105,15 @@ var SocketBase = /** @class */ (function (_super) {
         });
         this.binds = [];
         this.endpoints.forEach(function (endpoint) {
-            endpoint.removeListener('attach', _this.attachEndpoint);
-            endpoint.removeListener('terminated', _this.endpointTerminated);
+            endpoint.removeListener('attach', _this.endpointAttached);
+            endpoint.removeListener('lost', _this.endpointLost);
+            endpoint.removeListener('terminated', _this.endpointTerminatedInternal);
             endpoint.removeListener('message', _this.xrecv);
-            endpoint.removeListener('hiccuped', _this.hiccuped);
+            endpoint.removeListener('hiccuped', _this.endpointRecovered);
             endpoint.close();
             lodash_1.pull(_this.endpoints, endpoint);
             _this.endpointTerminated(endpoint);
+            _this.emit('lost', endpoint);
         });
     };
     SocketBase.prototype.subscribe = function (topic) {
@@ -108,11 +126,13 @@ var SocketBase = /** @class */ (function (_super) {
         endpoint.on('terminated', this.bindEndpointTerminated);
         endpoint.on('message', this.xrecv);
         this.attachEndpoint(endpoint);
+        this.emit('ready', endpoint);
     };
     SocketBase.prototype.bindEndpointTerminated = function (endpoint) {
         endpoint.removeListener('terminated', this.bindEndpointTerminated);
         endpoint.removeListener('message', this.xrecv);
         this.endpointTerminated(endpoint);
+        this.emit('lost', endpoint);
     };
     SocketBase.prototype.attachEndpoint = function (endpoint) {
     };
diff --git a/lib/src/webSocketEndpoint.js b/lib/src/webSocketEndpoint.js
index e9a18091fe3556507a4c80d160e2b154f0d93df0..9bbf0479bf3be7ca9865fd9a2ab379546cd4073e 100644
--- a/lib/src/webSocketEndpoint.js
+++ b/lib/src/webSocketEndpoint.js
@@ -48,6 +48,7 @@ var WebSocketEndpoint = /** @class */ (function (_super) {
             _this.accepted = true;
             _this.state = State.Active;
             _this.socket.binaryType = "arraybuffer";
+            _this.socket.onerror = _this.error.bind(_this);
             _this.socket.onclose = _this.onClose.bind(_this);
             _this.socket.onmessage = _this.onMessage.bind(_this);
             _this.send([_this.options.routingId]);
@@ -61,6 +62,7 @@ var WebSocketEndpoint = /** @class */ (function (_super) {
         this.socket = new WebSocket(this.address, ['ZWS2.0']);
         this.socket.binaryType = "arraybuffer";
         this.socket.onopen = this.onOpen.bind(this);
+        this.socket.onerror = this.error.bind(this);
         this.socket.onclose = this.onClose.bind(this);
         this.socket.onmessage = this.onMessage.bind(this);
     };
@@ -71,10 +73,10 @@ var WebSocketEndpoint = /** @class */ (function (_super) {
         this.send([this.options.routingId]);
         this.queue.forEach(function (frame) { return _this.socket.send(frame); });
         this.queue = [];
-        if (this.options.immediate)
-            this.emit('attach', this);
-        else if (oldState === State.Reconnecting)
+        if (oldState === State.Reconnecting)
             this.emit('hiccuped', this);
+        else
+            this.emit('attach', this);
     };
     WebSocketEndpoint.prototype.onClose = function () {
         if (this.accepted) {
@@ -82,8 +84,8 @@ var WebSocketEndpoint = /** @class */ (function (_super) {
             this.emit('terminated', this);
         }
         else if (this.state !== State.Closed) {
-            if ((this.state === State.Active || this.state === State.Connecting) && this.options.immediate)
-                this.emit('terminated', this);
+            if (this.state === State.Active || this.state === State.Connecting)
+                this.emit('lost', this);
             if (this.state === State.Active)
                 this.state = State.Reconnecting;
             setTimeout(this.connect, this.options.reconnectInterval);
diff --git a/src/socketBase.ts b/src/socketBase.ts
index fa4dd41f79a0cbb1541b18cf253b81cf0f09706f..900186f258511e2c9058d97fa3db256daa9cf4ed 100644
--- a/src/socketBase.ts
+++ b/src/socketBase.ts
@@ -12,6 +12,10 @@ class SocketBase extends EventEmitter {
     private endpoints: IEndpoint[] = []
     private binds: IListener[] = []
     public readonly options = new SocketOptions()
+    private readonly endpointAttached: (endpoint: IEndpoint) => void
+    private readonly endpointLost: (endpoint: IEndpoint) => void
+    private readonly endpointTerminatedInternal: (endpoint: IEndpoint) => void
+    private readonly endpointRecovered: (endpoint: IEndpoint) => void
 
     constructor() {
         super()
@@ -21,19 +25,32 @@ class SocketBase extends EventEmitter {
         this.endpointTerminated = this.endpointTerminated.bind(this)
         this.xrecv = this.xrecv.bind(this)
         this.hiccuped = this.hiccuped.bind(this)
+        this.endpointAttached = (endpoint: IEndpoint) => {
+            this.attachEndpoint(endpoint)
+            this.emit('ready', endpoint)
+        }
+        this.endpointLost = (endpoint: IEndpoint) => {
+            this.emit('lost', endpoint)
+        }
+        this.endpointTerminatedInternal = (endpoint: IEndpoint) => {
+            this.endpointTerminated(endpoint)
+            this.emit('lost', endpoint)
+        }
+        this.endpointRecovered = (endpoint: IEndpoint) => {
+            this.hiccuped(endpoint)
+            this.emit('ready', endpoint)
+        }
     }
 
     connect(address: string) {
         if (address.startsWith("ws://") || address.startsWith("wss://")) {
             const endpoint = new WebSocketEndpoint(address, this.options)
-            endpoint.on('attach', this.attachEndpoint)
-            endpoint.on('terminated', this.endpointTerminated)
+            endpoint.on('attach', this.endpointAttached)
+            endpoint.on('lost', this.endpointLost)
+            endpoint.on('terminated', this.endpointTerminatedInternal)
             endpoint.on('message', this.xrecv)
-            endpoint.on('hiccuped', this.hiccuped)
+            endpoint.on('hiccuped', this.endpointRecovered)
             this.endpoints.push(endpoint)
-
-            if (!this.options.immediate)
-                this.attachEndpoint(endpoint)
         } else {
             throw new Error('unsupported transport')
         }
@@ -43,13 +60,15 @@ class SocketBase extends EventEmitter {
         const endpoint = find(this.endpoints, e => e.address === address)
 
         if (endpoint) {
-            endpoint.removeListener('attach', this.attachEndpoint)
-            endpoint.removeListener('terminated', this.endpointTerminated)
+            endpoint.removeListener('attach', this.endpointAttached)
+            endpoint.removeListener('lost', this.endpointLost)
+            endpoint.removeListener('terminated', this.endpointTerminatedInternal)
             endpoint.removeListener('message', this.xrecv)
-            endpoint.removeListener('hiccuped', this.hiccuped)
+            endpoint.removeListener('hiccuped', this.endpointRecovered)
             endpoint.close()
             pull(this.endpoints, endpoint)
             this.endpointTerminated(endpoint)
+            this.emit('lost', endpoint)
         }
     }
 
@@ -86,13 +105,15 @@ class SocketBase extends EventEmitter {
         this.binds = []
 
         this.endpoints.forEach(endpoint => {
-            endpoint.removeListener('attach', this.attachEndpoint)
-            endpoint.removeListener('terminated', this.endpointTerminated)
+            endpoint.removeListener('attach', this.endpointAttached)
+            endpoint.removeListener('lost', this.endpointLost)
+            endpoint.removeListener('terminated', this.endpointTerminatedInternal)
             endpoint.removeListener('message', this.xrecv)
-            endpoint.removeListener('hiccuped', this.hiccuped)
+            endpoint.removeListener('hiccuped', this.endpointRecovered)
             endpoint.close()
             pull(this.endpoints, endpoint)
             this.endpointTerminated(endpoint)
+            this.emit('lost', endpoint)
         })
     }
 
@@ -109,6 +130,7 @@ class SocketBase extends EventEmitter {
         endpoint.on('message', this.xrecv)
 
         this.attachEndpoint(endpoint)
+        this.emit('ready', endpoint)
     }
 
     private bindEndpointTerminated(endpoint: IEndpoint) {
@@ -116,6 +138,7 @@ class SocketBase extends EventEmitter {
         endpoint.removeListener('message', this.xrecv)
 
         this.endpointTerminated(endpoint)
+        this.emit('lost', endpoint)
     }
 
     protected attachEndpoint(endpoint: IEndpoint) {
diff --git a/src/webSocketEndpoint.ts b/src/webSocketEndpoint.ts
index 2d3e225df5024bec410dea9e8866335a2cecb9d9..89c4f2ed65f60df122f0b34823499e401a098908 100644
--- a/src/webSocketEndpoint.ts
+++ b/src/webSocketEndpoint.ts
@@ -42,6 +42,7 @@ export default class WebSocketEndpoint extends EventEmitter implements IEndpoint
             this.accepted = true
             this.state = State.Active
             this.socket.binaryType = "arraybuffer"
+            this.socket.onerror = this.error.bind(this)
             this.socket.onclose = this.onClose.bind(this)
             this.socket.onmessage = this.onMessage.bind(this)
             this.send([this.options.routingId])
@@ -56,6 +57,7 @@ export default class WebSocketEndpoint extends EventEmitter implements IEndpoint
         this.socket = new WebSocket(this.address, ['ZWS2.0'])
         this.socket.binaryType = "arraybuffer"
         this.socket.onopen = this.onOpen.bind(this)
+        this.socket.onerror = this.error.bind(this)
         this.socket.onclose = this.onClose.bind(this)
         this.socket.onmessage = this.onMessage.bind(this)
     }
@@ -68,10 +70,10 @@ export default class WebSocketEndpoint extends EventEmitter implements IEndpoint
         this.queue.forEach(frame => this.socket.send(frame))
         this.queue = []
 
-        if (this.options.immediate)
-            this.emit('attach', this)
-        else if (oldState === State.Reconnecting)
+        if (oldState === State.Reconnecting)
             this.emit('hiccuped', this)
+        else
+            this.emit('attach', this)
     }
 
     onClose() {
@@ -80,8 +82,8 @@ export default class WebSocketEndpoint extends EventEmitter implements IEndpoint
             this.emit('terminated', this)
         }
         else if (this.state !== State.Closed) {
-            if ((this.state === State.Active || this.state === State.Connecting) && this.options.immediate)
-                this.emit('terminated', this)
+            if (this.state === State.Active || this.state === State.Connecting)
+                this.emit('lost', this)
 
             if (this.state === State.Active)
                 this.state = State.Reconnecting
